custom/icds_reports/management/commands/dashboard_usage_report.py

```diff
@@ -0,0 +1,156 @@
+
+prefix = 'dashboard_usage_data_'
+REQUEST_DATA_CACHE = f'{prefix}tabular_data.csv'
+CAS_DATA_CACHE = f'{prefix}cas_export_data.csv'
+
+
+def cache_to_file(cache_name):
+    def _outer(fn):
+        @wraps(fn)
+        def _inner(*args, **kwargs):
+            data = _get_from_file(cache_name)
+            if not data:
+                data = fn(*args, **kwargs)
+                _write_to_file(cache_name, data)
+            return data
+        return _inner
+    return _outer
+
+
+def _get_from_file(filename):
+    if os.path.exists(filename):
+        print(f'Fetching data from file: {filename}')
+        with open(filename, 'r') as f:
+            reader = csv.reader(f)
+            return [
+                r[0] if len(r) == 1 else r for r in list(reader)
+            ]
+
+
+def _write_to_file(filename, rows):
+    print(f'Writing {len(rows)} to file {filename}')
+    with open(filename, 'w') as f:
+        writer = csv.writer(f)
+        writer.writerows([
+            r if isinstance(r, list) else [r] for r in rows
+        ])
+
+
+class Command(BaseCommand):
+    def add_arguments(self, parser):
+        parser.add_argument(
+            'start_date',
+            type=date_type,
+            help='The start date (inclusive). format YYYY-MM-DD'
+        )
+        parser.add_argument(
+            'end_date',
+            type=date_type,
+            help='The end date (exclusive). format YYYY-MM-DD'
+        )
+        parser.add_argument('domain')
+        parser.add_argument('username', help='The username of the logged in user')
+
+    def handle(self, start_date, end_date, domain, username, **options):
+        print(f'fetching users data')
+
+        user = CouchUser.get_by_username(username)
+        usage_data = DashBoardUsage(couch_user=user, domain=domain).get_excel_data()[0][1]
+
+        # fetching dashboard usage counts
+        tab_usage_counts, tab_indicators_count, cas_counts = \
+            self.get_dashboard_usage_counts(start_date, end_date, domain)
+
+        username_state_mapping = self.get_tabular_data(usage_data, tab_usage_counts, tab_indicators_count)
+        print(f'Request data written to file {REQUEST_DATA_CACHE}')
+        self.get_cas_data(cas_counts, username_state_mapping)
+        print(f'Request data written to file {CAS_DATA_CACHE}')
+
+    def get_dashboard_usage_counts(self, start_date, end_date, domain):
+        """
+        :param start_date: start date of the filter
+        :param end_date: end date of the filter
+        :param domain
+        :return: returns the counts of no of downloads of each and total reports for  all usernames
+        """
+        print(f'Compiling usage counts for users')
+        tabular_user_counts = defaultdict(int)
+        cas_user_counts = defaultdict(int)
+        tabular_user_indicators = defaultdict(lambda: [0] * 10)
+
+        urls = ['/a/' + domain + '/cas_export', '/a/' + domain + '/icds_export_indicator']
```

Comment:
you could easily use f-strings here. No?

Reply:
since we are separting the two calls. we might not need this. I will change it to f string